internal/auth/internal/types/traffic_permissions.go

```diff
@@ -0,0 +1,101 @@
+package types
+
+import (
+	"github.com/hashicorp/go-multierror"
+
+	"github.com/hashicorp/consul/internal/resource"
+	pbauth "github.com/hashicorp/consul/proto-public/pbauth/v1alpha1"
+	"github.com/hashicorp/consul/proto-public/pbresource"
+)
+
+const (
+	NamespaceTrafficPermissionKind = "NamespaceTrafficPermission"
+	PartitionTrafficPermissionKind = "PartitionTrafficPermission"
+)
+
+var (
+	NamespaceTrafficPermissionV1AlphaType = &pbresource.Type{
+		Group:        GroupName,
+		GroupVersion: VersionV1Alpha1,
+		Kind:         NamespaceTrafficPermissionKind,
+	}
+
+	PartitionTrafficPermissionV1AlphaType = &pbresource.Type{
+		Group:        GroupName,
+		GroupVersion: VersionV1Alpha1,
+		Kind:         PartitionTrafficPermissionKind,
+	}
+
+	NamespaceTrafficPermissionType = NamespaceTrafficPermissionV1AlphaType
+	PartitionTrafficPermissionType = PartitionTrafficPermissionV1AlphaType
+)
+
+func RegisterTrafficPermissions(r resource.Registry) {
+	r.Register(resource.Registration{
+		Type:     NamespaceTrafficPermissionV1AlphaType,
+		Proto:    &pbauth.TrafficPermission{},
+		Scope:    resource.ScopeNamespace,
+		Validate: ValidateNamespaceTrafficPermission,
+	})
+	r.Register(resource.Registration{
+		Type:     PartitionTrafficPermissionV1AlphaType,
+		Proto:    &pbauth.PartitionTrafficPermission{},
+		Scope:    resource.ScopePartition,
+		Validate: ValidatePartitionTrafficPermission,
+	})
+}
+
+func ValidatePartitionTrafficPermission(res *pbresource.Resource) error {
+	return errNotSupported
+}
+
+func ValidateNamespaceTrafficPermission(res *pbresource.Resource) error {
+	var tp pbauth.TrafficPermission
+
+	if err := res.Data.UnmarshalTo(&tp); err != nil {
+		return resource.NewErrDataParse(&tp, err)
+	}
+
+	var err error
+
+	// TODO: Refactor errors for nested fields after further validation logic is added
+	if tp.Action == pbauth.Action_ACTION_UNSPECIFIED {
```

Comment:
this leaves options open for future action types beyond just "Allow" and "Deny", with more graceful compatibility: https://protobuf.dev/programming-guides/dos-donts/#unspecified-enum

Reply:
If we were to remove `UNSPECIFIED` now or later, I'd want to make the default "deny" so in the case of a bug where the field is left unset, permissions are overly restrictive rather than unenforced (security hole).

I was wondering the same as @ishustava though; it seems in this case the potential value would be more about forcing one of the two options to be set explicitly.

Reply:
>this leaves options open for future action types beyond just "Allow" and "Deny", with more graceful compatibility

I don't think we need that. Envoy has more action types, but it similarly defaults to `allow` (which is the default deny option Michael is talking about)